pkgs/development/compilers/dmd/default.nix

```diff
@@ -1,199 +1,163 @@
-      '';
-
-      installPhase = ''
-          echo -n $inputString > $out
-      '';
-    }
+  buildInputs = [ curl tzdata ];
+
+  bits = builtins.toString stdenv.hostPlatform.parsed.cpu.bits;
+  osname = if stdenv.hostPlatform.isDarwin then
+    "osx"
+  else
+    stdenv.hostPlatform.parsed.kernel.name;
+  top = "$(echo $NIX_BUILD_TOP)";
+  pathToDmd = "${top}/dmd/generated/${osname}/release/${bits}/dmd";
+
+  curlExtension = if stdenv.hostPlatform.isDarwin then
+    ".4.dylib"
+  else
+    ".so";
+
+  # Buid and install are based on http://wiki.dlang.org/Building_DMD
+  buildPhase = ''
+      cd dmd
+      make -j$NIX_BUILD_CORES -f posix.mak INSTALL_DIR=$out BUILD=release ENABLE_RELEASE=1 PIC=1 HOST_DMD=ldmd2
+      cd ../druntime
+      make -j$NIX_BUILD_CORES -f posix.mak BUILD=release ENABLE_RELEASE=1 PIC=1 INSTALL_DIR=$out DMD=${pathToDmd}
+      cd ../phobos
+      echo ${tzdata}/share/zoneinfo/ > TZDatabaseDirFile
+      echo ${curl.out}/lib/libcurl${curlExtension} > LibcurlPathFile
+      make -j$NIX_BUILD_CORES -f posix.mak BUILD=release ENABLE_RELEASE=1 PIC=1 INSTALL_DIR=$out DMD=${pathToDmd} DFLAGS="-version=TZDatabaseDir -version=LibcurlPath -J$(pwd)"
+      cd ..
+  '';
+
+  doCheck = true;
+
+  removeSocketTest = if stdenv.hostPlatform.isDarwin then
+    "rm phobos/std/socket.d"
   else
     "";
 
-in
+  checkPhase = ''
+    ${removeSocketTest}
 
-stdenv.mkDerivation rec {
-  inherit phobosUnittests;
-  name = "dmd-${version}";
-  phases = "installPhase";
-  buildInputs = dmdBuild.buildInputs;
+    cd druntime
+    make -j$NIX_BUILD_CORES -f posix.mak unittest PIC=1 DMD=${pathToDmd} BUILD=release
+    cd ../phobos
+    make -j$NIX_BUILD_CORES -f posix.mak unittest BUILD=release ENABLE_RELEASE=1 PIC=1 DMD=${pathToDmd} DFLAGS="-version=TZDatabaseDir -version=LibcurlPath -J$(pwd)"
+    cd ../dmd
+    make -j$NIX_BUILD_CORES -C test -f Makefile PIC=1 CC=$CXX DMD=${pathToDmd} BUILD=release SHELL=$SHELL
+    cd ..
+  '';
+
+  dontStrip = true;
 
   installPhase = ''
-    mkdir $out
-    cp -r --symbolic-link ${dmdBuild}/* $out/
+      cd dmd
+      mkdir $out
+      mkdir $out/bin
+      cp ${pathToDmd} $out/bin
+
+      mkdir -p $out/share/man/man1
+      mkdir -p $out/share/man/man5
+      cp -r docs/man/man1/* $out/share/man/man1/
+      cp -r docs/man/man5/* $out/share/man/man5/
+
+      cd ../druntime
+      mkdir $out/include
+      mkdir $out/include/d2
+      cp -r import/* $out/include/d2
+
+      cd ../phobos
+      mkdir $out/lib
+      cp generated/${osname}/release/${bits}/libphobos2.* $out/lib
+
+      cp -r std $out/include/d2
+      cp -r etc $out/include/d2
+
+      wrapProgram $out/bin/dmd \
+          --prefix PATH ":" "${targetPackages.stdenv.cc}/bin" \
+          --set-default CC "${targetPackages.stdenv.cc}/bin/cc"
+
+      cd $out/bin
+      tee dmd.conf << EOF
```

Comment:
Why not use `writeTextFile` and `lib.generators.toINI`?

Reply:
How is that better?

Reply:
I'd say it's just a better practice, since it's in nix for this purpose.